pallets/parachain-staking/src/inflation.rs

```diff
@@ -58,21 +60,30 @@ pub struct StakingInfo {
 }
 
 impl StakingInfo {
-	pub fn new(max_rate: Perquintill, annual_reward_rate: Perquintill) -> Self {
+	pub fn new<T: Config>(max_rate: Perquintill, annual_reward_rate: Perquintill) -> Self {
 		StakingInfo {
 			max_rate,
-			reward_rate: RewardRate::new(annual_reward_rate),
+			reward_rate: RewardRate::new::<T>(annual_reward_rate),
 		}
 	}
 
 	/// Calculate newly minted rewards on coinbase, e.g.,
 	/// reward = rewards_per_block * staking_rate.
-	pub fn compute_block_rewards<T: Config>(&self, stake: BalanceOf<T>, total_issuance: BalanceOf<T>) -> BalanceOf<T> {
-		let staking_rate = Perquintill::from_rational(stake, total_issuance).min(self.max_rate);
-		// multiplication with perbill cannot overflow
-		let rewards = self.reward_rate.per_block * total_issuance;
+	///
+	/// NOTE: If we exceed the max staking rate, the reward will be reduced by
+	/// max_rate / current_rate.
+	pub fn compute_reward<T: Config>(
+		&self,
+		stake: BalanceOf<T>,
+		current_staking_rate: Perquintill,
+		authors_per_round: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		// Perquintill automatically bounds to [0, 100]% in case staking_rate >
+		// self.max_rate
+		let reduction = Perquintill::from_rational(self.max_rate.deconstruct(), current_staking_rate.deconstruct());
 		// multiplication with perbill cannot overflow
-		staking_rate * rewards
+		let reward = self.reward_rate.per_block * stake * authors_per_round;
```

Comment:
`stake * authors per round` could overflow though?

sounds unlikely, but not free